[WIP] Add a new mixin for seeding

moolitayer opened the pull request:
## Goal

Streamline seeding. Currently seeding classes each define a data model (mostly for relationships with other objects) and models seeded together are nested in the same yml and have code for parsing in one of the classes. Nesting does not fit all cases since [examples from #7558]: 
- conditions can belong to more then one policy
- policies can belong to more then one policySets
- policies don't have to actually belong to a policySet 
- conditions don't have to belong to a policy)

Advantages of current approach and relying on guid for relationships:
- each model is seeded separately at model.seed (with order defined at evm_database.rb)
- no need to design different yml structure and write custom parsing code for models
## Usage

From #7558 ;

``` ruby
# Condition
seed_model

# from db/fixtures/conditions.yml

---
- :name: b0868f00-f525-11e5-9643-0242a1565ed5
  :description: Does not have cve failures with high severity
  :modifier: allow
  :expression: !ruby/object:MiqExpression
       exp:
         FIND:
           search:
             =:
               field: ContainerImage.openscap_result.openscap_rule_results-result
               value: fail
           checkall:
             REGULAR EXPRESSION DOES NOT MATCH:
               field: ContainerImage.openscap_result.openscap_rule_results-severity
               value: High
       context_type:
  :towhat: ContainerImage
  :file_mtime:
  :created_on:
  :updated_on:
  :guid: b0868f00-f525-11e5-9643-0242a1565ed5
  :filename:
  :applies_to_exp:
  :notes:

# MiqPolicy
seed_model(
  with_many(:conditions)
)

# from db/fixtures/miq_policies.yml

---
- :name: c5ceab70-da38-11e5-9cae-0242782a2608
  :description: Analyze all new images
  :expression:
  :towhat: ContainerImage
  :guid: c5ceab70-da38-11e5-9cae-0242782a2608
  :created_by: admin
  :updated_by: admin
  :notes:
  :active: true
  :mode: control
  :conditions: []

# MiqPolicyContent
def self.seed
  seed_model(
    with_one(:miq_policy),
    with_one(:miq_action, :name),
    with_one(:miq_event_definition),
  )
end

# from db/fixtures/miq_policy_contents.yml

---
- :guid: 116d2cd0-e50b-4e91-a094-6a6e64c079f3
  :miq_policy: c5ceab70-da38-11e5-9cae-0242782a2608
  :miq_action: container_image_analyze
  :miq_event_definition: containerimage_created
  :qualifier: failure
  :success_sequence:
  :failure_sequence: 1
  :success_synchronous:
  :failure_synchronous:

# MiqPolicySet
seed_model(
  with_many(:miq_policies, :guid, -> (policy_set, policy) { policy_set.add_member(policy) })
)

# from db/fixtures/miq_policy_sets.yml

---
- :name: 747a70d0-a8bf-4b63-bdad-1676bc0c5356
  :description: Container Image default profile
  :guid: 747a70d0-a8bf-4b63-bdad-1676bc0c5356
  :miq_policies:
    - c5ceab70-da38-11e5-9cae-0242782a2608

```

moolitayer commented:
ping @simon3z @jrafanie @gtanzillo @gmcculloug i

moolitayer commented on db/fixtures/conditions.yml:
```diff
@@ -0,0 +1,20 @@
+  :guid: 3366fc29-d457-4c64-a648-e2a5686600cd
+  :filename:
+  :applies_to_exp:
+  :miq_policy_guid: c5ceab70-da38-11e5-9cae-0242782a2608
```
I'm not sure why a condition has a miq_policy_id but i'm pretty sure I instead I need to seed conditions_miq_policies one to many association instead.

jrafanie commented on db/fixtures/miq_policy_contents.yml:
```diff
@@ -0,0 +1,9 @@
+---
+- :miq_policy_guid: c5ceab70-da38-11e5-9cae-0242782a2608
```
I think we shouldn't set all of these *guid via the fixture.

moolitayer replied:
@jrafanie I use the guids to link between related objects

moolitayer commented:
@miq-bot remove_label wip
@miq-bot add_label control

moolitayer commented:
Now seeding as MiqPolicySets (AKA profiles) as well. This will allow to attach defaults policy data to entities completely out of the box

simon3z commented:
@moolitayer at first glance I have the feeling that there's a lot of copy and paste. Any way of refactoring a little bit before proceeding?

jrafanie commented:
@moolitayer I like the general idea but I agree with @simon3z, it feels rather complicated and similar code.  Perhaps, start by creating methods to YAML.load_file + build file name and extract some of the similar behavior?  What do you think?

moolitayer commented:
Adapt logic to extract seed logic.
- remove relationship file in favor on inlining
- added code at evm_databases to time the seeding and divide to classes
- add seeding mixin so that seeding logic can be shared
- add specs for all seeds

moolitayer commented:
@jrafanie @Fryguy @gtanzillo This patch needs your review

gtanzillo commented:
This is really nice. Creating a `SeedingMixin` is something that should have been done a long time ago. I know that you have tests for this in #7558 but they seem to only test the creation of the seeded data (unless I missed it) but, it would be good to also test the update path as well.

jrafanie commented on app/models/miq_policy_set.rb:
```diff
@@ -116,6 +118,10 @@ def self.import_from_yaml(fd)
 
   def self.seed
+    seed_model(
+      self, :has_many => [:miq_policies, -> (policy_set, policy) { policy_set.add_member(policy) }]
```
@moolitayer First of all, this is awesome!  I'd like to see tests for this too... this api isn't immediately obvious to me, but I don't have useful suggestions so I'm fine with it... but maybe we can refine this api after we have some tests.

jrafanie commented on app/models/mixins/seeding_mixin.rb:
```diff
@@ -0,0 +1,51 @@
+      filter = options[:find_by] || -> (hash) { {:guid => hash[:guid]} }
+
+      load_fixtures(model.name.tableize).each do |hash|
+        association = options[:has_many].nil? ? nil : options[:has_many][0]
```
Maybe? 
```ruby
association, add_proc = options[:has_many]
```

moolitayer replied:
Thanks @jrafanie I wan't aware this is possible, (~~~a,b = nil~~~)

jrafanie commented on app/models/mixins/seeding_mixin.rb:
```diff
@@ -0,0 +1,51 @@
+      load_fixtures(model.name.tableize).each do |hash|
+        association = options[:has_many].nil? ? nil : options[:has_many][0]
+        rec = seed_attributes!(model, hash.except(association), filter)
+        seed_relationships(rec, hash, options) if options[:has_many]
```
I'd prefer you pass down the association and add_proc (or a better name for it) instead of an options hash.

jrafanie commented on app/models/mixins/seeding_mixin.rb:
```diff
@@ -0,0 +1,51 @@
+    def seed_relationships(rec, hash, options)
+      assoc = options[:has_many][0]
+      assoc_class = assoc.to_s.classify.constantize
+      assoc_method = options[:has_many][1]
```
If we pass the association name and the add_proc as above...

```ruby

def seed_relationships(rec, hash, association, add_proc)
  klass = association.to_s.classify.constantize
  hash[association].each do |guid|
    item = klass.find_by_guid(guid)
    add_proc.call(rec, klass.find_by_guid(guid)) unless item.nil?
  end
end
```

jrafanie replied:
We might want to pass along the actual association object in the future so we don't have to constantize.

moolitayer commented:
- Fixed according to comments from @jrafanie.
- Added tests but waiting before I test every flow (api needs to improve)

At #7567

@jrafanie wrote:
> Note, I'm wondering if the guid was because we need to seed separate classes and needed a way to link them up together... maybe if we seed them together in a single class, we can avoid needing this guid entirely?

@gtanzillo wrote:
>@moolitayer Regarding the guids - we added them to allow for creating policies (and other things that may not be policies) in one database, exporting them and importing them in another DB. And, updating them if they already exist in the destination DB. This was necessary because you can't rely on the ids being the same for the same policy in both DBs. So, the guid ensures that we can match them.

>This was a common pattern that was used in the field so that policies, some that were very complex, can be built and shared. 

Indeed in other models seeding multiple classes I've looked at, the yml is always nested. I believe that does not fit this kind of data: conditions can belong to more then one policy, same goes for policies and policySets. In addition policies don't have to actually belong to a policySet (same goes for conditions and policies) so I'm not sure nesting is any good.

I understand currently seeding guid isn't something that we do. Question is, is there a reason not to?
If I have a guid for each associated entity we gain:
- each model is  seeded separately at model.seed (with order defined at evm_database.rb)
- no need to design different yml structure and custom parsing code

gtanzillo commented:
@moolitayer 

> I understand currently seeding guids isn't something that we do. Question is, is there a reason not to?
If I have a guid for each associated entity we gain:

each model is seeded separately at model.seed (with order defined at evm_database.rb)
no need to design different yml structure and custom parsing code for models

I agree with . I'd seed the guid column with a guid and the name column with an actual name instead of the guid.

moolitayer commented:
@jrafanie There are some Issues I see with my current implementation, maybe you know how to improve it;
- Maybe I need to remove the :search_filter_for_hash that knows how to get the filter from yml hash (e.g for guid: ```-> (hash) { {:guid => hash[:guid]} }```) and add a guid to policyContent?
- Same model, It might not make sense to do [this](https://github.com/ManageIQ/manageiq/pull/7413/files#diff-99b0453f9aee19a8261e7ece7a179b57R24) merge for future users. I use it to add found event and action to policyContent. Maybe I should replace it with has_one ?
- Maybe I need to not be strict on the guid and allow listing other fields in associations?

This is the Best API I can think of (usage example):
```
Condition:
seed_model(self)

MiqPolicy:
seed_model(
  self,
  :has_many  => [
    {
      :name     => :conditions,
      :column   => :guid,
      :add_proc => -> (policy, condition) { policy.conditions << condition }
    }
  ]
)

MiqPolicyContent:
seed_model(
  self,
  :has_one => [
    {
      :name     => :miq_policy
      :column   => :guid,
    },
    {
      :name     => :miq_event_definition,
      :column   => :name,
    },
    {
      :name     => :miq_action,
      :column   => :name,
    },
  ]
)

MiqPolicySet:
seed_model(
  self,
  :has_many => [
    {
      :name      => :miq_policies,
      :column    => :guid,
      :add_proc  => -> (policy_set, policy) { policy_set.add_member(policy) }
    }
  ]
)

```
That would require adding guid to policyContent to drop the :search_filter_for_hash option.

@jrafanie @gtanzillo sorry for all the text, Waiting for input before I proceed

jrafanie commented:
@moolitayer this part feels like a "struct" or something:

```ruby
  :has_many => [
    {
      :name      => :miq_policies,
      :column    => :guid,
      :add_proc  => -> (policy_set, policy) { policy_set.add_member(policy) }
    }
```

Maybe something like this?

```ruby
AssociationSeed = Struct.new(:association, :type, :find_by, :add_proc)
association_seed = AssociationSeed.new(:miq_policies, :has_many, :guid, -> (policy_set, policy) { policy_set.add_member(policy)})

# MiqPolicySet:
seed_model(self, [association_seed])
```

simon3z commented:
@moolitayer who should be reviewing this? @jrafanie ? Anyone else?

moolitayer commented:
> @moolitayer who should be reviewing this? @jrafanie ? Anyone else?

@jrafanie and @gtanzillo started reviewing #7413 but they are unavailable this week so definitely in need of more reviewers, maybe @blomquisg has someone in mind.

@bdunne @durandom @Fryguy could you take a look at this? Thanks

chrisarcand commented:
@moolitayer Could you add a detailed description summarizing the goal of all of this (especially for those just now looking at it?)

This seems to be adding an entire DSL for seeding on top of what we already have. Have we considered SeedFu or some other tested thing and utilizing guids which would make primordial seeding fairly trivial and simple (hopefully)?

moolitayer commented:
> @moolitayer Could you add a detailed description summarizing the goal of all of this (especially for those just now looking at it?)

added goal section to first pr comment, see also comments [1] (https://github.com/ManageIQ/manageiq/pull/7413#issuecomment-202883148),[2](https://github.com/ManageIQ/manageiq/pull/7413#issuecomment-203202108)


> This seems to be adding an entire DSL for seeding on top of what we already have. Have we considered SeedFu or some other tested thing and utilizing guids which would make primordial seeding fairly trivial and simple (hopefully)?

If you have any other suggestions for seeding I can use for #7558 instead of this it might be a good idea.
#7558 needs to get in and this change was requested (comments 1 and 2 mentioned above) and had [initial] agreement (I admit I do not know SeedFu or any other seeding libs)  

Note this has many tests I was requested for [here](https://github.com/moolitayer/manageiq/blob/predefined_policies/spec/models/mixins/seeding_mixin_spec.rb) and 4 more [here](https://github.com/ManageIQ/manageiq/pull/7558/files)

ps Utilizing guid is exactly what I do as discussed [here](https://github.com/ManageIQ/manageiq/pull/7413#issuecomment-206441243) and agreed upon [here](https://github.com/ManageIQ/manageiq/pull/7413#issuecomment-206449133)

chrisarcand commented:
I would very much suggest forgoing adding a custom DSL to this and instead utilizing simple find_or_creates or [SeedFu's constraint feature](https://github.com/mbleigh/seed-fu#constraints) with GUIDs.

I'm honestly in favor of a simple PORO (with or without seedfu) and removing the `::seed` methods on all the classes, centralizing everything. The way we seed currently is scattered all over the app, order-dependent, confusing, etc, and to be honest I feel like this adds a bunch of complexity on top of something already needlessly complex 😕 I/we just haven't gotten around to overhauling it. 

However, all that I just mentioned is well beyond the scope of what you immediately might need. If you need to get something seeded immediately I would suggest just patching in what you need with the current setup. It should be possible, you just can't put the seed logic for every related many relationship completely separate in each model. For example, I need to seed the many entitlements per miq_group in miq_group (nothing in entitlements). 

All my opinion of course, I'm curious what other people might think of this direction. @jrafanie @Fryguy ?

durandom commented:
> I'm honestly in favor of a simple PORO ... centralizing everything. 

👍  here. In my experience utilizing the least amount of magic and plain ActiveRecord objects is the most descriptive way of doing seeding. It also helps discovering design flaws in the models sometimes.

@moolitayer I know this does not help you with getting #7558 in though :( Do you have an older commit where you seed just with the `def self.seed` method?

Fryguy commented:
I agree with @chrisarcand.  We definitely need to clean up seeding, so the general direction is good, but the approach here seems unnecessarily complex.  I'm not sure it really "fixes" the problem, it just obscures it behind a DSL.

I don't like the addition of the guid.  Creating a proper guid will probably not be done by a developer when they add a new entry.  Instead they are likely to just take the previous one and change the last digit, which is probably not a good thing.  Instead, I think it would be nicer if the caller could pass what are the unique columns, for example `seed_model(:name)` if the name is what's considered unique.

> If you need to get something seeded immediately I would suggest just patching in what you need with the current setup.

Mixing the immediate need to seed certain tables in #7558 with a complete refactoring here, seems like a bad approach, particularly since #7558 will get held up until this PR is done.  #7558 should just get the data in the current form to solve that need, and this PR should probably change to fix all (or at least many) of the seeding classes at once.


I need the opinions of @chessbyte and @jrafanie and others, but right now I am not a fan of the approach here.

simon3z commented:
> Mixing the immediate need to seed certain tables in #7558 with a complete refactoring here, seems like a bad approach, particularly since #7558 will get held up until this PR is done.

@Fryguy I agree since this PR is becoming controversial.

@moolitayer do you think you can bake up something in #7558 to get it merged and postpone this refactoring?

moolitayer commented:
Marking this as wip for now. Seeding of policies moved to #7966
@miq-bot add_label wip

moolitayer commented:
This does not seem like the way to go, and no resources/need for it right now
